cpp/src/arrow/python/iterators.h

```diff
@@ -97,31 +98,53 @@ inline Status VisitSequence(PyObject* obj, int64_t offset, VisitorFunc&& func) {
 template <class VisitorFunc>
 inline Status VisitSequenceMasked(PyObject* obj, PyObject* mo, int64_t offset,
                                   VisitorFunc&& func) {
-  if (mo == nullptr || !PyArray_Check(mo)) {
-    return Status::Invalid("Null mask must be NumPy array");
-  }
-
-  PyArrayObject* mask = reinterpret_cast<PyArrayObject*>(mo);
-  if (PyArray_NDIM(mask) != 1) {
-    return Status::Invalid("Mask must be 1D array");
-  }
+  if (PyArray_Check(mo)) {
+    PyArrayObject* mask = reinterpret_cast<PyArrayObject*>(mo);
+    if (PyArray_NDIM(mask) != 1) {
+      return Status::Invalid("Mask must be 1D array");
+    }
+    if (PyArray_SIZE(mask) != static_cast<int64_t>(PySequence_Size(obj))) {
+      return Status::Invalid("Mask was a different length from sequence being converted");
+    }
 
-  const Py_ssize_t obj_size = PySequence_Size(obj);
-  if (PyArray_SIZE(mask) != static_cast<int64_t>(obj_size)) {
-    return Status::Invalid("Mask was a different length from sequence being converted");
-  }
+    const int dtype = fix_numpy_type_num(PyArray_DESCR(mask)->type_num);
+    if (dtype == NPY_BOOL) {
+      Ndarray1DIndexer<uint8_t> mask_values(mask);
 
-  const int dtype = fix_numpy_type_num(PyArray_DESCR(mask)->type_num);
-  if (dtype == NPY_BOOL) {
-    Ndarray1DIndexer<uint8_t> mask_values(mask);
+      return VisitSequenceGeneric(
+          obj, offset,
+          [&func, &mask_values](PyObject* value, int64_t i, bool* keep_going) {
+            return func(value, mask_values[i], keep_going);
+          });
+    } else {
+      return Status::Invalid("Mask must be boolean dtype");
+    }
+  } else if (PySequence_Check(mo)) {
```

Comment:
If we would like to restrict this branch to pyarrow arrays, then you could try to 
1. `unwrap_array(mo)` to access the underlying C++ array
2. then try to downcast it to BooleanArray
3. access the boolean value of the C++ array directly rather than going through the python scalar API